sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -114,18 +181,67 @@ RT::PiKernel ProgramManager::getOrCreateKernel(OSModuleHandle M,
     std::cerr << ">>> ProgramManager::getOrCreateKernel(" << M << ", "
               << getRawSyclObjImpl(Context) << ", " << KernelName << ")\n";
   }
-  RT::PiProgram Program = getBuiltOpenCLProgram(M, Context);
+
+  auto Program = getBuiltOpenCLProgram(M, Context);
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
-  std::map<RT::PiProgram, std::map<string_class, RT::PiKernel>> &CachedKernels =
-      Ctx->getCachedKernels();
-  std::map<string_class, RT::PiKernel> &KernelsCache = CachedKernels[Program];
-  RT::PiKernel &Kernel = KernelsCache[KernelName];
-  if (!Kernel) {
-    PI_CALL(piKernelCreate)(Program, KernelName.c_str(), &Kernel);
+
+  using PiKernelT = KernelProgramCache::PiKernelT;
+  using KernelWithBuildState = KernelProgramCache::KernelWithBuildStateT;
+
+  KernelProgramCache &Cache = Ctx->getKernelProgramCache();
+  bool InsertionTookPlace;
+  KernelWithBuildState *KernelWithState;
+
+  {
+    auto Locked = Cache.acquireKernelsPerProgramCache();
+    auto &KernelsPerProgram = Locked.get();
+    auto &KernelsCache = KernelsPerProgram[Program];
+
+    auto Inserted = KernelsCache.emplace(
+        std::piecewise_construct, std::forward_as_tuple(KernelName),
+        std::forward_as_tuple(nullptr, BS_InProgres));
+
+    InsertionTookPlace = Inserted.second;
+    KernelWithState = &Inserted.first->second;
+  }
+
+  // any thread which will find nullptr in cache will wait until the pointer
+  // is not null anymore
+  if (!InsertionTookPlace) {
+    return waitUntilBuilt<invalid_object_error>(Cache, KernelWithState);
+  }
+
+  // we may now build the kernel whilst other threads are waiting
+  try {
+    PiKernelT *Desired = nullptr;
+
     // TODO need some user-friendly error/exception
     // instead of currently obscure one
+    PI_CALL(piKernelCreate)(Program, KernelName.c_str(), &Desired);
+
+    PiKernelT *Expected = nullptr;
+
+#ifndef NDEBUG
+    if (!KernelWithState->Ptr.compare_exchange_strong(Expected, Desired)) {
+      // We've got a funny story here
+      assert(false && "We've build a kernel that is already have been built.");
+    }
+#else
+    KernelWithState->Ptr.store(Desired);
+#endif
+
+    KernelWithState->State.store(BS_Done);
+
+    Cache.notifyAllBuild();
+
+    return Desired;
+  } catch (...) {
+    KernelWithState->State.store(BS_Failed);
+
+    Cache.notifyAllBuild();
+
+    std::rethrow_exception(std::current_exception());
```

Comment:
`throw;` ?

All this code looks déjà vu... Is there some possible reasonable refactoring?
Perhaps some generic caching code taking some lambdas precising some actions for kernels or programs?
Just some raw guess...